### DISCUSSION TITLE
Alias / scope `x-data` in the DOM via attribute `:value`.

### DISCUSSION THREAD

**Discussion by bennadel** [alpinejs/alpine#4053] [3. Feature ideas]
I'm still learning Alpine.js, so maybe this isn't even a thing worth considering. But, the fact that we can essentially nest `x-data` components means that it can be unclear where a particular property is being read from (and possibly written to?). I wonder if there would be value in being able to alias an `x-data` controller in the dom using the directive "value". Imagine this:

```html
<div x-data:outer="{ value: 1 }">
	<div x-data:inner="{ value: 1 }">

		<button @click="outer.value++">
			Increment Outer Value
		</button>
		<button @click="inner.value++">
			Increment Inner Value
		</button>

	</div>
</div>
```

By using `x-data:outer`, it allows me to specify _which_ reactive scope I'm actually trying to reference, as in `outer.value++` vs `inner.value++`.

I know that you could likely achieve the same type of thing by specifying `outer` and `inner` as root properties in the reactive structure:

`x-data="{ outer: { value: 1 } }"`

But, this means that said data structure is now more tightly coupled to any methods defined on that component as well.

Maybe this already exists and I just haven't seen it in the documentation :D 

Anyway, just some thoughts as an Alpine.js n00b.

**Comment by bennadel** [comment]
I am borrowing this idea from Angular.js, which feels like it has a lot of overlap with Alpine. In Angular, you can alias "controllers" like:

```html
<div ng-controller="MyController as vm">
    <button ng-click="vm.doSomething()"> ... </button>
</div>
```

They use `as {alias}` as their syntax. At least I think - I haven't used this technique in a while (they have other ways of aliasing stuff that is more config-based).

**Comment by ekwoka** [comment]
I think just nesting the data would make more sense if it was needed.

Broadly I'd question what the purpose is of such shadowing.

It might be an XY problem

**Comment by bennadel** [comment]
@ekwoka sorry, I think perhaps I mislead with my question. I have no interest in actually nesting two different properties named `value` - I only put that out there as a way to illustrate the means of differentiating one reactive scope from another. I figure there are two types of use-cases: one where it's needed in order to work (such as above) and one where it needed only for clarity (ie, to clearly identify which scope manages which variable).

In AngularJS, the scopes work very similar to AlpineJS. And, I can only say that over time, scope inheritance became a point of frustration for people because they had no idea where data was coming from. I'm only trying to cross-pollinate some thinking here :) 

Many of the Alpine example involve people hard-coding a lot of logic right into the `x-data` attribute. But, consider the CSP version, where everything becomes labels. In that case, it might be nice to have a way to differentiate:

```html
<div x-data:x="outer">
	<div x-data:y="inner">
		<button @click="x.doThis">
			Click This
		</button>
		<button @click="y.doThat">
			Click That
		</button>
	</div>
</div>
```

Anyway, I'm just trying to share some reflections as someone who is learning Alpine.js. 

Also, what is an "XY Problem" - I'm not familiar with that term.

**Reply by ekwoka** [reply]
An XY problem is a problem that someone is facing that is because they chose some way to do something at some other point where what they did there was the real issue.

It can come up a lot in these kinds of open help questions where people ask "how do I do X?" And the X seems really strange and difficult to do. And over time you find out the real problem was Y that they thought X would solve, but a fundamentally different approach to Y solves it better to where X never becomes an issue.

Like in the presented example with 'value' being shadowed, I'd be concerned about why value would need to be shadowed in the first place.

But yes, some kind of scoping of the data could be nice...

Maybe if `$data` would expose keys for the id on an `x-data` element provided a closer context doesn't overwrite it...

Or an explicit name...

Curious.... :thinking:

I can see ways to do it fairly simply, but making the case for it to exist....I'm not sure if I'm just too used to it....

**Reply by bennadel** [reply]
Ah, that makes sense - thanks for the clarity. I totally can relate to that type of problem!

I don't quite have a good sense of how Alpine is architected under the hood - I've only poked around at some of the existing directives. I gather it walks up some sort of "data stack parent path". If that's the case, you could probably just add the label as a property on the current scope object. Then, the "label" in the DOM is really just a "property" on the scope - it's just that it was added automatically by the framework.

Pseudo code for `x-data`:

```js
function XData( element, metadata, framework ) {
    if ( metadata.value ) {
        this[ metadata.value ] = this; // Assign scope back to itself using given directive property.
    }
}
```

But again, I'm just trying to cross-pollinate ideas. If this isn't something Alpine'ites aren't asking for, then I don't intent to add more work :)

**Comment by bennadel** [comment]
For anyone that happens upon this discussion, I created a proof-of-concept using a sibling directive, `x-scope`:

```html
<div x-data="{ value: 0 }" x-scope="me">
    <span x-text="me.value"></span>
</div>
```

Here's my exploration: https://www.bennadel.com/blog/4598-scoping-reactive-object-references-in-the-dom-in-alpine-js-3.htm

**Reply by mattvague** [reply]
Hey @bennadel, thanks for mentioning this. Having clashing `x-data` properties is a problem I've also been running into a lot lately especially when building reusable plugins.

@ekwoka I'd like to put my support in for some sort of way of scoping data (whether it's by something like `x-scope="myScope"` or by adding a modifier to `x-data` like `x-data.myScope="..."` or something else). Alpine is amazing, but I do feel like the current design severely limits plugin flexibility and/or makes plugin development more onerous than they need to be.

For example, let's say I have a (very simplified) modal and a (very simplified) disclosure plugin, both with a `__isOpen` property:

```
// modal.js
Alpine.directive('modal', ($el) => {
  Alpine.bind($el, {
    "x-data"() {
      return {
        __isOpen: true
      }
    },
    "x-show": '__isOpen'
  });
})

Alpine.magic('modal', ($el) => {
  const $data = Alpine.$data($el)
  return {
    get isOpen() { return $data.__isOpen },
    set isOpen(open) { $data.__isOpen = open }
  }
})

```

```
// disclosure.js
Alpine.directive('disclosure', ($el) => {
  Alpine.bind($el, {
    "x-data"() {
      return {
        __isOpen: true
      }
    },
    "x-show": '__isOpen'
  });
})

Alpine.magic('disclosure', ($el) => {
  const $data = Alpine.$data($el)
  return {
    get isOpen() { return $data.__isOpen },
    set isOpen(open) { $data.__isOpen = open }
  }
})
```

Then, If for some reason I use my disclosure within my modal like this:

```
<div x-data x-modal>
    My Modal!
    <button @click="$modal.isOpen = false">Close Modal 1</button>

    <div x-disclosure>
      <button @click="$modal.isOpen = false">Close Modal 2</button>
      <button @click="$disclosure.isOpen = false">Close Disclosure</button>
    </div>
  </div>
```

When I click the "Close Modal 2" button it will actually close the disclosure, not the modal since the magics are both naively manipulating the closest property called `__isOpen`. See a [full gist here](https://gist.github.com/mattvague/9997a1106aaf66e7dcdd21044cc573e4).

The way I've been getting around this is to namespace all plugin x-data functions and properties, but this is pretty ugly and tedious:

```
// disclosure.js
Alpine.directive('disclosure', ($el) => {
  Alpine.bind($el, {
    "x-data"() {
      return {
        __disclosure_isOpen: true
      }
    },
    "x-show": '__disclosure_isOpen'
  });
})

Alpine.magic('disclosure', ($el) => {
  const $data = Alpine.$data($el)
  return {
    get isOpen() { return $data. __disclosure_isOpen },
    set isOpen(open) { $data. __disclosure_isOpen = open }
  }
})
```

@ekwoka or @calebporzio is there any appetite for trying to solve this problem? I'd be happy to try out a few possibilities in a PR (maybe working in concert with @bennadel)

**Reply by bennadel** [reply]
I'm always happy to help out, but I'm not that advanced on this stuff yet :) I didn't even know there was an `Apline.$data(el)` function - is that doing something similar to `Alpine.closestDataStack(el)`. I also haven't used any magics yet (except for refs). It's funny, the API is rather small, but the concepts are powerful and take a bit to sink in.

But, you paint a good picture here. Essentially, any "generic" sounding thing like "open" "close" "toggle" "active", there's a non-zero chance that conflicts can arise. Good discussion!

**Reply by mattvague** [reply]
>I didn't even know there was an Apline.$data(el) function

Yeah that's not documented, I found it by reading through the [Alpine UI plugin code](https://github.com/alpinejs/alpine/tree/main/packages/ui/src). I think that's an actually a pretty good way to understand how @calebporzio is intending things to be used and pick up on some tips and tricks. Lots of undocumented stuff there

> is that doing something similar to Alpine.closestDataStack(el)

Yes, more or less. I believe this is the implementation of `Alpine.$data`: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/scope.js#L2-L4. 

> I also haven't used any magics yet (except for refs). It's funny, the API is rather small, but the concepts are powerful and take a bit to sink in.

Yeah it's super powerful once you understand how the reactive data, directives, and magics all work together. I feel like this kind of "advanced" usage could be documented a bit better. Also happy to help out with this

**Reply by bennadel** [reply]
@mattvague totally off-topic now, but agreed that some of this stuff just needs to be stumbled-upon. I keep looking at the source code for inspiration in how to build my own experiments - but the source code is quite different since it has access to all the internal modules. Right now, I'm just trying to think of fun things to build that force me to understand the framework better. I honestly forgot that there was even a UI part in the docs 😆 right now, just trying to drink from the fire-host. Thanks for that link, I think that will prove quite valuable.

**Reply by ekwoka** [reply]
Woah, I just did this this morning

https://www.npmjs.com/package/@ekwoka/alpine-scope

As a proof of concept. Will be interesting to compare approaches

**Reply by ekwoka** [reply]
@mattvague I think your above example with your plugins can be solved better with using symbols to just scope your relevant behaviors and not inject stuff directly on the x-data context in such a fragile way.

If you look at my above alpine-scope source you can see it stores the context info with a symbol, so they can't really squash.

[Here's an example in action](https://awesomealpine.com/play?coreplugins=0&html=PGRpdiB4LWRhdGE9InsgdmFsdWU6IG51bGwgfSI-CiAgPGRpdiB4LWRpc2Nsb3N1cmU-CiAgICA8YnV0dG9uIEBjbGljaz0iJGRpc2Nsb3N1cmUuaXNPcGVuID0gZmFsc2UiPgogICAgICBDbG9zZSBEaXNjbG9zdXJlCiAgICA8L2J1dHRvbj4KICA8L2Rpdj4KPC9kaXY-Cg&script=Ly8gZGlzY2xvc3VyZS5qcwpjb25zdCAkZGlzY2xvc3VyZSA9IFN5bWJvbCgpOwpBbHBpbmUuZGlyZWN0aXZlKCdkaXNjbG9zdXJlJywgKGVsKSA9PiB7CiAgY29uc3QgY29udGV4dCA9IEFscGluZS5yZWFjdGl2ZSh7CiAgICBpc09wZW46IHRydWUsCiAgfSk7CiAgQWxwaW5lLmFkZFNjb3BlVG9Ob2RlKGVsLCB7IFskZGlzY2xvc3VyZV06IGNvbnRleHQgfSk7CiAgQWxwaW5lLmJpbmQoZWwsIHsKICAgICd4LXNob3cnKCkgewogICAgICByZXR1cm4gY29udGV4dC5pc09wZW47CiAgICB9LAogIH0pOwp9KS5iZWZvcmUoJ2JpbmQnKTsKCkFscGluZS5tYWdpYygnZGlzY2xvc3VyZScsIChlbCkgPT4gewogIGNvbnN0IGNvbnRleHQgPSBBbHBpbmUuJGRhdGEoZWwpWyRkaXNjbG9zdXJlXTsKICBpZiAoIWNvbnRleHQpCiAgICB0aHJvdyBuZXcgRXJyb3IoCiAgICAgICckZGlzY2xvc3VyZSBuZWVkcyB0byBiZSB1c2VkIGluc2lkZSBhbiB4LWRpc2Nsb3N1cmUgY29tcG9uZW50JywKICAgICk7CiAgcmV0dXJuIGNvbnRleHQ7Cn0pOwo&panel=script)

```js
// disclosure.js
const $disclosure = Symbol();
Alpine.directive('disclosure', (el) => {
  const context = Alpine.reactive({
    isOpen: true,
  });
  Alpine.addScopeToNode(el, { [$disclosure]: context });
  Alpine.bind(el, {
    'x-show'() {
      return context.isOpen;
    },
  });
}).before('bind');

Alpine.magic('disclosure', (el) => {
  console.log(JSON.stringify(Alpine.$data(el), null, 2));
  const context = Alpine.$data(el)[$disclosure];
  if (!context)
    throw new Error(
      '$disclosure needs to be used inside an x-disclosure component',
    );
  return context;
});
```

**Reply by ekwoka** [reply]
On the note about undocumented stuff, if you add `@types/alpinejs` you will get some extra documentation on some of these methods, like `Alpine.$data` that will show when you hover over them. I've been adding JSDoc to the type definitions.

```ts
{
    /**
     * Gets the data context of a particular Node
     * This is a "flattened" object of all the components that element is within
     * @param node Element inside an Alpine Component
     * @returns Object
     */
    $data: (node: ElementWithXAttributes) => {};
}
```

**Reply by bennadel** [reply]
@ekwoka so, when looking at this code in your plugin:

```js
const context = Alpine.$data(el) as Scopable;
const rootContext = Alpine.closestDataStack(el)[0];
```

Are you saying that `context` here is is a flattened object that contains all the scopes? And the `rootContext` is the scope for the given element only? I'm still trying to get up-to-speed on how all the internal methods work, so this is non-obvious to me. It's cool to look at your implementation, though.

**Reply by bennadel** [reply]
Wow, reading through the internal `mergeProxyTrap` code is so fascinating! I've only dabbled with Proxy objects in the past; this whole Scope concept brings it to a new level.

**Reply by ekwoka** [reply]
Yes, context is the merged proxy that accesses the entire data stack. The same one used in the expressions. That's the `mergeProxies` function.

I wrote the current version of the proxy merging logic. The other version Caleb wrote had mostly same functionality, but this version [reduced memory and increased performance](https://github.com/alpinejs/alpine/pull/3722). The PR might help make the concepts clearer.

When you access a property, it first checks the nearest for that key, then the next, then the next, then the next, and then magics normally.

It returns the first it finds.

The most "local" state wins.

So children shadow parents values but don't squash them.

**Reply by bennadel** [reply]
It's very clever. I wanted to put together a [small demo for myself](https://www.bennadel.com/blog/4600-overcoming-asymmetric-prototype-property-access-with-proxies-in-javascript.htm). Digging into the internals of a library can be so inspiring.

**Reply by mattvague** [reply]
> @mattvague I think your above example with your plugins can be solved better with using symbols to just scope your relevant behaviors and not inject stuff directly on the x-data context in such a fragile way.

> If you look at my above alpine-scope source you can see it stores the context info with a symbol, so they can't really squash.

Whoaaaaaa! That is crazy cool, thank you so much. 

>On the note about undocumented stuff, if you add @types/alpinejs you will get some extra documentation on some of these methods, like Alpine.$data that will show when you hover over them. I've been adding JSDoc to the type definitions.

Also thank you for maintaining the TS types. I have some ideas on how to improve those as well, will try and contribute soon.

**Reply by ekwoka** [reply]
you and me both!
